### PR TITLE
IGNITE-20416 Retry when schema change is detected during implicit transaction

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-20416

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java [hunk 1]
```diff
@@ -163,11 +164,14 @@ private <T> CompletableFuture<T> withSchemaSync(@Nullable Transaction tx, @Nulla
...
 
                                 // Repeat.
+                                return withSchemaSync(tx, schemaVersion, action);
+                            } else if (tx == null && isOrCausedBy(IncompatibleSchemaException.class, ex)) {
+                                // Schema has changed while we were executing an implicit transaction: let's retry.
+                                assertSchemaVersionIncreased(previousSchemaVersion, schemaVersion);
+
                                 return withSchemaSync(tx, schemaVersion, action);
                             } else {
                                 return ex == null ? completedFuture(res) : CompletableFuture.<T>failedFuture(ex);
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java:177]: I would move the null check at the very beginning, it's a bit more clear: we first check the normal outcome and then start checking the exceptions

File: modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java [hunk 2]
```diff
@@ -163,11 +164,14 @@ private <T> CompletableFuture<T> withSchemaSync(@Nullable Transaction tx, @Nulla
...
                             if (isOrCausedBy(InternalSchemaVersionMismatchException.class, ex)) {
                                 assert tx == null : "Only for implicit transactions a retry might be requested";
-                                assert previousSchemaVersion == null || !Objects.equals(schemaVersion, previousSchemaVersion)
-                                        : "Same schema version (" + schemaVersion
-                                        + ") on a retry: something is wrong, is this caused by the test setup?";
+                                assertSchemaVersionIncreased(previousSchemaVersion, schemaVersion);
 
                                 // Repeat.
+                                return withSchemaSync(tx, schemaVersion, action);
+                            } else if (tx == null && isOrCausedBy(IncompatibleSchemaException.class, ex)) {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java:171]: If this represents a schema change, then was does `InternalSchemaVersionMismatchException` do?

File: modules/platforms/dotnet/Apache.Ignite.Tests/Table/SchemaSynchronizationTest.cs [hunk 3]
```diff
@@ -339,7 +339,6 @@ public async Task TestClientUsesLatestSchemaOnReadKv()
         [Values(true, false)] bool insertNewColumn,
         [Values(true, false)] bool withRemove)
     {
-        using var metricListener = new MetricsTests.Listener();
```
Comment (sashapolo) [inline modules/platforms/dotnet/Apache.Ignite.Tests/Table/SchemaSynchronizationTest.cs:342]: What's this about?

Comment (rpuch) [reply modules/platforms/dotnet/Apache.Ignite.Tests/Table/SchemaSynchronizationTest.cs:342]: This was added to work around a transaction abort that could happen due to a concurrent schema change and an implicit transaction (generated by the streamer)

File: modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java (hunk 2 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java:171]: When doing a KV operation and we DON'T have an explicit transaction, this is what happens:

1. We take the 'current' table version at 'now' and marshal keys/values using this version (this is moment T1)
2. Then we call a method of `InternalTable`, it creates a new transaction at moment T2
3. Then an operation is executed at moment T3

`InternalSchemaVersionMismatchException` is for a case when table version at T2 is different from table version at T1. This is needed to close that gap between marshalling and starting a transaction: to make sure that transaction works with rows in the version it expects, but avoid creating a transaction in the KV view (we tried this, it causes troubles)

And `IncompatibleSchemaException` is for a case when table version at T3 is different from table version at T2.

Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java:175]: Can we add corresponding comments inside the branches describing these situations?

**Review by sashapolo — APPROVED** [review]

File: modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java (hunk 2 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java:181]: Added comments

**Pull request merged**